---
title: "Differential gene expression analysis of mutated regulators"
author: "Bingru Sun"
date: "2018/11/21"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,warning = FALSE)
library(limma)
library(org.Hs.eg.db)
library(annotate)
library(GEOquery)
library(grid)
library(ggplot2)
library(GO.db)
library(GSVA)
library(GSEABase)
library(data.table)
```

```{r regulator mutation input,include = FALSE}
## input of mutation data
mutation_o<-read.delim('./data/CCLE/CCLE_DepMap_18Q1_maf_20180207 (1).txt')
mutation<-mutation_o
mutation_s<-mutation_o[which(is.na(mutation_o$ExAC_AF)),]
mutation_s<-mutation_s[which(mutation_s$isDeleterious == 'TRUE'),]
mutation_s_w<-mutation_s
mutation_s_w<-data.frame(mutation_s$Hugo_Symbol,mutation_s$Tumor_Sample_Barcode)

## put mutated cell lines into a list
genelist<-as.character(unique(mutation_s_w[,1]))
rs<-function(x){
  c<-as.character(mutation_s_w[which(mutation_s_w$mutation_s.Hugo_Symbol==x),2])
  return(c)
  }
a<-lapply(genelist,rs)
names(a)<-genelist
mutatedcelllines<-a
b<-lapply(mutatedcelllines,length)
mutatedcelllines_num<-unlist(b)

## get regulator target gene set 
regset <- getGmt('./data/Gene_regulators/targetset_nodup.gmt.txt')
reg.list<-geneIds(regset)
reg.count<-lapply(reg.list,length)
reg.count<-unlist(reg.count)
  
regulator<-names(reg.count)
target <- unique(unlist(reg.list))

## put the regulator target gene set into a matrix
reg_m<-matrix(nrow=length(regulator),ncol=length(target))
colnames(reg_m)<-target
row.names(reg_m)<-regulator


for (i in 1:nrow(reg_m)){
  x <- regulator[i]
  for (y in reg.list[[i]])
  {reg_m[x,y] <- 1}
}

reg_m[is.na(reg_m)]<-0

```

```{r ccle data input}
## selected 770 cell lines that have information of mutations, CNAs and expression with only hematopoeitic cell lines

load('./data/selected_cl.Rdata')

## read expression data
ccle.expression<-fread('./data/CCLE/CCLE_Expression_Entrez_2012-09-29.gct',skip = 2, sep = '\t')
ccle.expression_o <- ccle.expression
ccle.expression <- as.data.frame(ccle.expression)
ccle.expression <- ccle.expression[,-c(1,2)]
ccle.anno <- ccle.expression_o[,c(1,2)]
toremove <- which(ccle.anno$Description == '')
ccle.anno <- ccle.anno[-toremove,]
ccle.expression<-ccle.expression[-toremove,]


sample.info<-fread('./data/CCLE/CCLE_sample_info_file_2012-10-18.txt', sep = '\t')

ccle.exp <- ccle.expression[,cl.tokeep]

cl.sample <- colnames(ccle.exp)
n <- length(cl.sample)
cl<-rep(0,n)

regulators.cl <- intersect(regulator, names(mutatedcelllines))

mutation_df <- sapply(regulators.cl,function(x){return( cl.sample %in% mutatedcelllines[[x]])})

row.names(mutation_df) <- cl.sample

mutation_df <- t(mutation_df)

mt.count <- rowSums(mutation_df)
```

```{r dge}
## function of doing dge for each regulator 
dge.mutation<-function(gene){
n<-length(cl.sample)
cl.design <- matrix(nrow=n,ncol=2)
cl.design[,1] <- as.numeric(as.logical(mutation_df[gene,]))
cl.design.2<-cl.design[,1]
cl.design.2[which(cl.design[,1] == 1)]<- 0
cl.design.2[which(cl.design[,1] == 0)]<- 1
cl.design[,2]<-cl.design.2
colnames(cl.design)<-c('mutated','non_mutated')
fit <- lmFit(ccle.exp, cl.design)
contrasts <- makeContrasts(mutated - non_mutated, levels=cl.design)
fit2 <- contrasts.fit(fit, contrasts)
fit2 <- eBayes(fit2)
fit2$genes <- ccle.anno$Description
topTable(fit2)
testResults <- topTable(fit2, number=nrow(fit2))
testResults <- testResults[complete.cases(testResults),]
return(as.character(testResults[which(testResults$ID == gene),] ))
}

regulators.mut <- regulators.cl[mt.count > 0]

regulator_dge_df <- sapply(regulators.mut,dge.mutation)

#save(regulator_dge_df,file = '../data/regulator_dge.Rdata')
```

##Method
Differential gene expression analysis is conducted using limma. The regulator expression in the mutated cell lines is compared versus the regulator expression in the non-mutated cell lines. The results of the significant differential expressed genes are listed below.     

##Results of significant differential expressed regulators 
*(logFC >= 1 or <= -1, adjusted p-value <= 0.05)
```{r data}
row.names(regulator_dge_df) <- c('ID','logFC','AveExpr','t','P.value','adj.P.Val','B')

regulator_dge_df <- as.data.frame(regulator_dge_df)

regulator_dge_df <- t(regulator_dge_df)

regulator_dge_tb <- data.table(regulator_dge_df)

regulator_dge_tb[,logFC := as.numeric(logFC)]

regulator_dge_tb[,adj.P.Val := as.numeric(adj.P.Val)]

#regulator_dge_tb[(logFC >= 1 | logFC <= -1) & adj.P.Val <= 0.05]

tb <- regulator_dge_tb[(logFC >= 1 | logFC <= -1) & adj.P.Val <= 0.05]

knitr::kable(tb)
```

##Results of significant differential expressed regulators 
*(logFC >= 1 or <= -1, adjusted p-value <= 0.05) of regulators mutated in at least 5 cell lines and 20 downstream target genes

```{r regulators }
## load selected regulators
load('./data/regulators_5_cl_20_tar.Rdata')
tb <- regulator_dge_tb[ID %in% selected.regulator & adj.P.Val <= 0.05 & (logFC >= 1 | logFC <= -1)]
knitr::kable(tb)
```

## results in lung cancer cell lines
```{r lung dge}
lung.cl<-cl.sample[grep('LUNG',cl.sample)]
mutation_df.lung <- mutation_df[,lung.cl]

dge.mutation.lung<-function(gene){
n<-length(lung.cl)
cl.design <- matrix(nrow=n,ncol=2)
cl.design[,1] <- as.numeric(as.logical(mutation_df.lung[gene,]))
cl.design.2<-cl.design[,1]
cl.design.2[which(cl.design[,1] == 1)]<- 0
cl.design.2[which(cl.design[,1] == 0)]<- 1
cl.design[,2]<-cl.design.2
colnames(cl.design)<-c('mutated','non_mutated')
fit <- lmFit(ccle.exp[,lung.cl], cl.design)
contrasts <- makeContrasts(mutated - non_mutated, levels=cl.design)
fit2 <- contrasts.fit(fit, contrasts)
fit2 <- eBayes(fit2)
fit2$genes <- ccle.anno$Description
topTable(fit2)
testResults <- topTable(fit2, number=nrow(fit2))
testResults <- testResults[complete.cases(testResults),]
return(as.character(testResults[which(testResults$ID == gene),] ))
}

lung.count <- rowSums(mutation_df.lung)

lung.mut <- regulators.cl[lung.count > 0]

regulator_dge_df.lung <- sapply(lung.mut,dge.mutation.lung)

row.names(regulator_dge_df.lung) <- c('ID','logFC','AveExpr','t','P.value','adj.P.Val','B')

regulator_dge_df.lung <- as.data.frame(regulator_dge_df.lung)

regulator_dge.lung <- t(regulator_dge_df.lung)

regulator_dge_tb.lung <- data.table(regulator_dge.lung)

regulator_dge_tb.lung[,logFC := as.numeric(logFC)]

regulator_dge_tb.lung[,adj.P.Val := as.numeric(adj.P.Val)]

tb <- regulator_dge_tb.lung[(logFC >= 1 | logFC <= -1) & adj.P.Val <= 0.05]

knitr::kable(tb)
```


## results in skin cancer cell lines
```{r skin age}
skin.cl<-cl.sample[grep('SKIN',cl.sample)]
mutation_df.skin <- mutation_df[,skin.cl]

dge.mutation.skin<-function(gene){
n<-length(skin.cl)
cl.design <- matrix(nrow=n,ncol=2)
cl.design[,1] <- as.numeric(as.logical(mutation_df.skin[gene,]))
cl.design.2<-cl.design[,1]
cl.design.2[which(cl.design[,1] == 1)]<- 0
cl.design.2[which(cl.design[,1] == 0)]<- 1
cl.design[,2]<-cl.design.2
colnames(cl.design)<-c('mutated','non_mutated')
fit <- lmFit(ccle.exp[,skin.cl], cl.design)
contrasts <- makeContrasts(mutated - non_mutated, levels=cl.design)
fit2 <- contrasts.fit(fit, contrasts)
fit2 <- eBayes(fit2)
fit2$genes <- ccle.anno$Description
topTable(fit2)
testResults <- topTable(fit2, number=nrow(fit2))
testResults <- testResults[complete.cases(testResults),]
return(as.character(testResults[which(testResults$ID == gene),] ))
}

skin.count <- rowSums(mutation_df.skin)

skin.mut <- regulators.cl[skin.count > 0]

regulator_dge_df.skin <- sapply(skin.mut,dge.mutation.skin)

row.names(regulator_dge_df.skin) <- c('ID','logFC','AveExpr','t','P.value','adj.P.Val','B')

regulator_dge_df.skin <- as.data.frame(regulator_dge_df.skin)

regulator_dge.skin <- t(regulator_dge_df.skin)

regulator_dge_tb.skin <- data.table(regulator_dge.skin)

regulator_dge_tb.skin[,logFC := as.numeric(logFC)]

regulator_dge_tb.skin[,adj.P.Val := as.numeric(adj.P.Val)]

tb <- regulator_dge_tb.skin[(logFC >= 1 | logFC <= -1) & adj.P.Val <= 0.05]

knitr::kable(tb)
```


```{r save results,eval = FALSE}
## save results
regulator.dge <- regulator_dge_tb[(logFC >= 1 | logFC <= -1) & adj.P.Val <= 0.05]$ID
regulator.lung.dge<- regulator_dge_tb.lung[(logFC >= 1 | logFC <= -1) & adj.P.Val <= 0.05]$ID
regulator.skin.dge<- regulator_dge_tb.skin[(logFC >= 1 | logFC <= -1) & adj.P.Val <= 0.05]$ID
save(regulator.dge ,regulator.lung.dge,regulator.skin.dge, file = "../data/dge.RData")

```

